Reduce rerendering on startup

sgomes opened the pull request:
## Description
Small tweaks that result in less unnecessary work during startup. This is a very modest speed improvement (after the changes, the time spent in JS during startup goes down by about 3-5%), but it doesn't introduce much additional complexity at just 2 lines of code.

## How has this been tested?
Unit tests, ad-hoc testing. Changes are minor and easy to analyse.

For performance measuring, I tried to create as stable an environment as possible on my local machine (with as few applications running as possible), throttled the CPU by 6x, and ran startup profiles for a large post with many blocks, repeatedly. For each of these runs, I measured how much time the sum of the two long frames during startup took.

![image](https://user-images.githubusercontent.com/409615/49952244-7e0ba300-fef3-11e8-8839-2ef018287fc6.png)

## Types of changes

**Ensure viewport calculations happen synchronously instead of waiting on event loop.**

We need to call `flush`, so that the changes happen synchronously, rather than asynchronously, on the next event loop. This avoids the re-rendering of all `BlockListBlock` instances during startup.

**Ensure `state.hasUploadPermissions` is always a boolean.**

The reducer for this property defaults to an empty object, which not only is incorrect (since this is a boolean), but also causes unnecessary re-renders by changing reference. By defaulting to `true` instead, we fix both problems.

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style.
- [X] My code follows the accessibility standards.
- [X] My code has proper inline documentation.

refs #11782

sgomes committed:
Ensure viewport data store changes happen immediately on startup.

We need to call `flush`, so that the changes happen synchronously,
rather than asynchronously, on the next event loop.
This avoids the re-rendering of all `BlockListBlock` instances
during startup.

sgomes commented:
Tests seem to be failing due to a configuration issue with `phpcs`.

aduth commented on packages/core-data/src/selectors.js:
```diff
@@ -178,5 +178,5 @@ export function isPreviewEmbedFallback( state, url ) {
  */
 export function hasUploadPermissions( state ) {
-	return state.hasUploadPermissions;
+	return !! state.hasUploadPermissions;
```
What's the reason for this change? This should be something where, if we're expecting it to be a boolean, it should be forced to be a boolean in the reducer logic, not derived by the selector.

sgomes replied:
Force-pushed this change. This is now handled at the reducer level where we default to `true` instead of `{}`. I chose `true` instead of `false`, since `{}` is a truthy value, and this should thus maintain existing behaviour.

aduth replied:
I'm guessing the previous implementation was a result of some bad copypasta from other reducers. A reducer value prefixed with "has" should almost certainly hold a boolean value.

Were there unit tests which should be updated as part of the change? Can there be, if not? 😄

sgomes replied:
Nope, this particular reducer didn't seem to have any unit tests. I added a couple :)

sgomes committed:
Ensure `state.hasUploadPermissions` is always a boolean.

The reducer for this property defaults to an empty object, which
not only is incorrect (since this is a boolean), but also causes
unnecessary re-renders by changing reference. By defaulting to
`true` instead, we fix both problems.

sgomes committed:
Add a couple of tests for hasUploadPermissions reducer.

aduth commented:
> * Ensure selector returns a primitive type instead of a changing object reference

Could you elaborate on why this change was made? I mean, it seems like a perfectly correct fix, just not sure how it relates to the performance work.

aduth commented:
I might expect it having something to do with the fact that state "changed" previously (from `{}` to `true`)? Which would cascade to subscribers (potentially many bound components)? Related reading if you're interested on the front of "oh wow we do a lot of state changes during initialization": #12877.

aduth committed:
Core Data: Add CHANGELOG entry for hasUplaodPermissions boolean fix

aduth committed:
Viewport: Add CHANGELOG entry for synchronous assignment

aduth reviewed (approved):
LGTM 👍 

I added some CHANGELOG entries for the affected packages, which we should be more in the habit of doing to inform the package publisher on the sorts of changes included (in selecting major, minor, patch).

sgomes commented:
Thanks for the link to #12877! That's a lot better than any startup optimisations opportunities I've found so far, which tended to have a nice large impact in development mode, but next to nothing in production. I assume this has to do with the time it takes to do the extra logging that takes place in development.

Also, thanks for adding the changelog commits!
